// gitHubEs/entities/pull-request-metadata.js
var pullRequestMetadata = {
        getCreateObject: function(data, updateData, config) {
            pr = {
                id: data.id,
                user_id: data.user.id,
                number: data.number,
                labels: [],
                title: data.title,
                body: data.body,
                created_at: data.created_at,
                updated_at: data.updated_at,
                closed_at: data.closed_at,
                merged_at: data.marged_at,
                merge_commit_sha: data.merge_commit_sha,
                assignee: data.assignee ? data.assignee.id : null,
                assignees: [],
                merged: data.merged,
                mergeable: data.mergeable,
                rebaseable: data.rebaseable,
                mergeable_state: data.mergeable_state,
                merged_by: data.merged_by,
                comments: data.comments,
                review_comments: data.review_comments,
                maintainer_can_modify: data.maintainer_can_modify,
                commits: data.commits,
                additions: data.additions,
                deletions: data.deletions,
                changed_files: data.changed_files,
                baseOrganisation: data.base.user.login,
                baseRepo: data.base.repo.name
            }
            updateObject  = this.getUpdateObject(pr, updateData, config)
            Object.keys(updateObject).map(function(key, index) {
                pr[key] = updateObject[key];
            });
            return pr
        },

        getUpdateObject: function(document, updateData, config) {
            updateObject = {}
            if (updateData.hasOwnProperty('labels')) {
                labels = document.hasOwnProperty('labels') ? document.labels : [];
                updateData.labels.map(function(label) {
                    if (!labels.find(function(element) {return element.id == label.id})) {
                        labels.push(updateData.label);
                        updateObject.labels = labels;
                    }
                })
            }
            //case 'unlabeled':
            return updateObject;
        }
}

module.exports = pullRequestMetadata;
